web/src/app/(main)/games/[slug]/goals/_components/GoalCodeDialog.tsx

```diff
@@ -0,0 +1,233 @@
+'use client';
+import {
+    Dialog,
+    DialogContent,
+    DialogTitle,
+    Button,
+    Box,
+    Typography,
+    IconButton,
+} from '@mui/material';
+import FullscreenIcon from '@mui/icons-material/Fullscreen';
+import FullscreenExitIcon from '@mui/icons-material/FullscreenExit';
+import { useState, useEffect, useCallback } from 'react';
+import { useGoalManagerContext } from '../../../../../../context/GoalManagerContext';
+import JsonEditor from './JsonEditor';
+import { notifyMessage } from '../../../../../../lib/Utils';
+
+interface GoalCodeDialogProps {
+    isOpen: boolean;
+    close: () => void;
+    slug: string;
+}
+
+export default function GoalCodeDialog({
+    isOpen,
+    close,
+    slug,
+}: GoalCodeDialogProps) {
+    const { goals, shownGoals, mutateGoals } = useGoalManagerContext();
+    const [jsonValue, setJsonValue] = useState('');
+    const [isLoading, setIsLoading] = useState(false);
+    const [error, setError] = useState<{
+        line?: number;
+        column?: number;
+        message?: string;
+    } | null>(null);
+    const [isFullscreen, setIsFullscreen] = useState(false);
+
+    // Determine the order to use: match UI ordering when it represents the full list
+    const getExportOrder = useCallback(() => {
+        if (shownGoals && shownGoals.length === goals.length) {
+            return shownGoals;
+        }
+        return goals;
+    }, [goals, shownGoals]);
+
+    // Convert goals to JSON format (dynamically include all fields except excluded ones)
+    const exportGoalsToJson = useCallback(() => {
+        const ordered = getExportOrder();
+        const goalsForExport = ordered.map((goal) => {
+            // Exclude system fields that shouldn't be editable
+            const excludedFields = new Set([
+                'id',
```

Comment:
Some of these properties exist on neither the database schema for goals nor the typedef, what is it referencing? Would it be easier/safer to implement this as an allowlist, rather than a blocklist since it's likely that the list of fields on the goal model will continue to grow, but it's unlikely that all of them should editable in this way